File: android/app/src/main/java/com/engchatgemini/com/voice/VoiceModule.java
package com.engchatgemini.com.voice;

import android.media.MediaRecorder;
import android.os.Build;
import android.os.Environment;
import android.util.Log;
import androidx.annotation.NonNull;
import com.facebook.react.bridge.*;

import java.io.File;
import java.io.IOException;

public class VoiceModule extends ReactContextBaseJavaModule {
    private static final String TAG = "VoiceModule";
    private String filePath;
    private MediaRecorder recorder;

    public VoiceModule(ReactApplicationContext reactContext) {
        super(reactContext);
    }

    @NonNull
    @Override
    public String getName() {
        return "VoiceModule";
    }

    @ReactMethod
    public void startRecording(Promise promise) {
        try {
            filePath = getReactApplicationContext().getExternalFilesDir(Environment.DIRECTORY_MUSIC).getAbsolutePath() + "/recording.mp3";
            recorder = new MediaRecorder();
            recorder.setAudioSource(MediaRecorder.AudioSource.MIC);
            recorder.setOutputFormat(MediaRecorder.OutputFormat.MPEG_4);
            recorder.setOutputFile(filePath);
            recorder.setAudioEncoder(MediaRecorder.AudioEncoder.AAC);
            recorder.prepare();
            recorder.start();
            promise.resolve(filePath);
        } catch (IOException e) {
            promise.reject("RECORDING_ERROR", "Lỗi khi ghi âm: " + e.getMessage());
        }
    }
    
    @ReactMethod
    public void stopRecording(Promise promise) {
        try {
            recorder.stop();
            recorder.release();
            recorder = null;
            promise.resolve(filePath);
        } catch (Exception e) {
            promise.reject("STOP_ERROR", "Lỗi khi dừng ghi âm: " + e.getMessage());
        }
    }
}

File: android/app/src/main/java/com/engchatgemini/com/voice/VoiceModule.kt
package com.engchatgemini.com.voice

import android.media.MediaPlayer
import android.util.Log
import com.facebook.react.bridge.*
import java.io.IOException

class VoiceModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
    private var mediaPlayer: MediaPlayer? = null
    private var currentUrl: String? = null

    override fun getName(): String {
        return "VoiceModule"
    }

    @ReactMethod
    fun playAudio(url: String, promise: Promise) {
        try {
            if (mediaPlayer == null) {
                mediaPlayer = MediaPlayer()
                mediaPlayer?.setOnCompletionListener {
                    Log.d("VoiceModule", "Playback completed")
                }
            } else {
                mediaPlayer?.reset()
            }

            mediaPlayer?.setDataSource(url)
            mediaPlayer?.prepareAsync()
            mediaPlayer?.setOnPreparedListener {
                it.start()
                currentUrl = url
                promise.resolve("Playing")
            }
        } catch (e: IOException) {
            promise.reject("PLAY_ERROR", "Lỗi khi phát âm thanh: ${e.message}")
        }
    }

    @ReactMethod
    fun pauseAudio(promise: Promise) {
        mediaPlayer?.let {
            if (it.isPlaying) {
                it.pause()
                promise.resolve("Paused")
            } else {
                it.start()
                promise.resolve("Resumed")
            }
        } ?: promise.reject("NO_PLAYER", "MediaPlayer chưa được khởi tạo")
    }

    @ReactMethod
    fun stopAudio(promise: Promise) {
        mediaPlayer?.let {
            it.stop()
            it.reset()
            promise.resolve("Stopped")
        } ?: promise.reject("NO_PLAYER", "MediaPlayer chưa được khởi tạo")
    }

    @ReactMethod
    fun seekForward(promise: Promise) {
        mediaPlayer?.let {
            val newPosition = it.currentPosition + 10_000
            it.seekTo(newPosition.coerceAtMost(it.duration))
            promise.resolve("Seeked Forward")
        } ?: promise.reject("NO_PLAYER", "MediaPlayer chưa được khởi tạo")
    }

    @ReactMethod
    fun seekBackward(promise: Promise) {
        mediaPlayer?.let {
            val newPosition = it.currentPosition - 10_000
            it.seekTo(newPosition.coerceAtLeast(0))
            promise.resolve("Seeked Backward")
        } ?: promise.reject("NO_PLAYER", "MediaPlayer chưa được khởi tạo")
    }
}
